### PR TITLE
Optimization Under Uncertainty

### PR DESCRIPTION
**Pull request by tomoleary**
Adding support for optimization problems, where there is an additional paramter `z` that is optimized in expectation over the distribution of parameter `m`, e.g., the prior

### CODE REVIEW

**Comment by tomoleary** [comment]
Should we add evaluation helper arrays to `Observable` classes in order to avoid constant need for new allocations?

File: hippyflow/modeling/activeSubspaceProjector.py [hunk 1]
```diff
@@ -544,34 +631,83 @@ def _construct_low_rank_Jacobians_serial(self,check_for_data = True,compress_fil
...
+			output_dimension,parameter_dimension = self.J.shape
+			parameter_rank = min(self.parameters['jacobian_rank'],output_dimension,parameter_dimension)
+
+		if parameter_dimension is None:
+			parameter_dimension = self.m.size()
+
+		if control_jacobian:
+			self.Jz = ObservableControlJacobian(self.observable)
+			output_dimension,control_dimension = self.Jz.shape
+			control_rank = min(self.parameters['jacobian_rank'],output_dimension,control_dimension)
```
Comment (dc-luo) [inline hippyflow/modeling/activeSubspaceProjector.py:666]: Should this be `self.parameters['control_jacobian_rank']`

File: hippyflow/modeling/observable.py [hunk 2]
```diff
@@ -114,14 +115,14 @@ def generate_vector(self, component = "ALL"):
...
-		elif component == STATE:
+		elif component == hp.STATE:
 			x = self.problem.generate_state()
-		elif component == PARAMETER:
+		elif component == hp.PARAMETER:
 			x = self.problem.generate_parameter()
-		elif component == ADJOINT:
+		elif component == hp.ADJOINT:
 			x = self.problem.generate_state()
 		elif component == 3:
```
Comment (dc-luo) [inline hippyflow/modeling/observable.py:124]: I think `CONTROL` is defined in this file. Let's use that instead of 3

**Review by dc-luo — CHANGES_REQUESTED** [review]

File: hippyflow/modeling/activeSubspaceProjector.py (hunk 1 above)
Comment (tomoleary) [reply hippyflow/modeling/activeSubspaceProjector.py]: Thanks!

**Review by dc-luo — APPROVED** [review]

**Pull request merged**